pkg/virt-controller/watch/vm_initializer_test.go

```diff
@@ -0,0 +1,440 @@
+package watch
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+)
+
+var _ = Describe("VM Initializer", func() {
```

Comment:
These are great tests that target each part of the initialization process.

One thing that is lacking is an end to end test of the initializeVirtualMachine() function. This is important to ensure the sanity of the entire flow.  Take a look at the vm_test.go tests. There's ways to mock the informers, cache, and client to ensure we're aware of any changes that occur to during the entire chain of events that occur during initializeVirtualMachine().

Reply:
The fact that there's no end-to-end tests of the entire thing is exactly why I've left this PR as a "work in progress". I don't consider it done without that sanity check. :)

Reply:
take a look at how the vm_test.go unit tests exercise the controller's Execute() function by using a mock api server.